mlflow/server/middleware.py

```diff
@@ -0,0 +1,57 @@
+from starlette.middleware import Middleware
+from starlette.requests import Request, HTTPConnection
+from starlette_context import plugins
+from starlette_context.middleware import RawContextMiddleware
+from typing import Optional, Union, Any
+
+
+class QueryParamsPlugin(plugins.base.Plugin):
+    key = "query_params"
+
+    async def process_request(
+        self, request: Union[Request, HTTPConnection]
+    ) -> Optional[Any]:
+        assert isinstance(self.key, str)
+        return request.query_params
+
+
+class BodyPlugin(plugins.base.Plugin):
+    key = "body"
+
+    async def process_request(
+        self, request: Union[Request, HTTPConnection]
+    ) -> Optional[Any]:
+        assert isinstance(self.key, str)
+        if hasattr(request, "body"):
+            return await request.body()
+        return None
+
+
+class MethodPlugin(plugins.base.Plugin):
+    key = "method"
+
+    async def process_request(
+        self, request: Union[Request, HTTPConnection]
+    ) -> Optional[Any]:
+        assert isinstance(self.key, str)
+        return request.method
+
+
+class RequestContextMiddleware(RawContextMiddleware):
+    @staticmethod
+    def get_request_object(
+        scope, receive, send
+    ) -> Union[Request, HTTPConnection]:
+        return Request(scope, receive, send)
+
+
+middleware = [
+    Middleware(
+        RequestContextMiddleware,
+        plugins=(
+            QueryParamsPlugin(),
+            BodyPlugin(),
+            MethodPlugin(),
+        )
```

Comment:
this middleware adds a globally available thread safe `context` variable that let us access the current request payload and headers, similar to Flask's `request`. 

`RequestContextMiddleware` intercepts the request object. The plugins define which attribute to put into `context.data` dictionary.

This requires an external package `starlette_context`, but the code is simple and we can also easily rewrite the middleware from scratch ourselves